java/vector/src/main/java/org/apache/arrow/vector/complex/ListViewVector.java

```diff
@@ -0,0 +1,913 @@
+   *
+   * @return the value capacity
+   */
+  @Override
+  public int getValueCapacity() {
+    return getValidityAndOffsetValueCapacity();
+  }
+
+  private int getValidityAndSizeValueCapacity() {
+    final int offsetValueCapacity = Math.max(getOffsetBufferValueCapacity(), 0);
+    final int sizeValueCapacity = Math.max(getSizeBufferValueCapacity(), 0);
+    return Math.min(offsetValueCapacity, sizeValueCapacity);
+  }
+
+  private int getValidityAndOffsetValueCapacity() {
+    final int offsetValueCapacity = Math.max(getOffsetBufferValueCapacity(), 0);
+    return Math.min(offsetValueCapacity, getValidityBufferValueCapacity());
+  }
+
+  private int getValidityBufferValueCapacity() {
+    return capAtMaxInt(validityBuffer.capacity() * 8);
+  }
+
+  /**
+   * Set the element at the given index to null.
+   * @param index the value to change
+   */
+  @Override
+  public void setNull(int index) {
+    while (index >= getValidityAndSizeValueCapacity()) {
+      reallocValidityAndSizeAndOffsetBuffers();
+    }
+    if (lastSet >= index) {
+      lastSet = index - 1;
+    }
+
+    if (index == 0) {
+      offsetBuffer.setInt(0, 0);
+      sizeBuffer.setInt(0, 0);
+    } else {
+      final int prevOffset = offsetBuffer.getInt((index - 1) * OFFSET_WIDTH);
+      final int prevSize = sizeBuffer.getInt((index - 1) * SIZE_WIDTH);
+      final int currOffSet = prevOffset + prevSize;
+      offsetBuffer.setInt(index * OFFSET_WIDTH, currOffSet);
+      sizeBuffer.setInt(index * SIZE_WIDTH, 0);
+    }
+
+    BitVectorHelper.unsetBit(validityBuffer, index);
+    lastSet = index;
+  }
+
+  /**
+   * Start new value in the ListView vector.
+   * There are a few cases that are handled in this function.
+   * There are two main scenarios that need to be considered.
+   * The first scenario is simple insertion where indices are continuously updated.
+   * The other scenario is the event of non-continuous writing,
+   * the offset buffer needs to be updated.
+   *
+   * @param index index of the value to start
+   * @return offset of the new value
+   */
+  @Override
+  public int startNewValue(int index) {
+    while (index >= getValidityAndSizeValueCapacity()) {
+      reallocValidityAndSizeAndOffsetBuffers();
+    }
+
+    if (lastSet >= index) {
+      lastSet = index - 1;
+    }
+
+    if (index == 0) {
+      offsetBuffer.setInt(0, 0);
+    } else if (index > lastSet) {
+      /* when skipping indices, we need to update the offset buffer */
+      /* setting offset from lastSet + 1 to index (included) */
+      for (int i = lastSet + 1; i <= index; i++) {
+        if (i == 0) {
+          offsetBuffer.setInt(0, 0);
+          continue;
+        }
+        final int prevOffSet = offsetBuffer.getInt((i - 1L) * OFFSET_WIDTH);
+        final int prevSize = sizeBuffer.getInt((i - 1L) * SIZE_WIDTH);
+        final int currOffSet = prevOffSet + prevSize;
+        offsetBuffer.setInt(i * OFFSET_WIDTH, currOffSet);
+      }
```

Comment:
I'm still not convinced we need this...we zero initialize the offset/size/validity buffers so this should automatically happen (any indices you don't set will just be null/0 length already).

Reply:
You mean the `if` clause on setting 0th index right?

Reply:
this whole block, and lastSet

Reply:
I decided to use lastSet for the case of keeping a flag to populate offsets when indices are skipped. So meaning we need to figure out a different alternative?

Reply:
But it doesn't matter right? We want a null list. The offset and size can just be 0.

Reply:
This was only needed to preserve the invariant for the normal list vector. That invariant doesn't exist here so why spend so much effort to preserve it?

Reply:
Let me rethink this. I get your point.

Reply:
@lidavidm I have a question what would be the offsets and sizes for the following according to your thought?

I have a doubt about 3rd index value in offset

```java
try (ListViewVector listViewVector = ListViewVector.empty("sourceVector", allocator)) {
      UnionListViewWriter listViewWriter = listViewVector.getWriter();

      /* allocate memory */
      listViewWriter.allocate();

      /* write the first list at index 0 */
      listViewWriter.setPosition(0);
      listViewWriter.startList();

      listViewWriter.bigInt().writeBigInt(12);
      listViewWriter.bigInt().writeBigInt(-7);
      listViewWriter.bigInt().writeBigInt(25);
      listViewWriter.endList();

      /* the second list at index 1 is null (we are not setting any)*/

      /* write the third list at index 2 */
      listViewWriter.setPosition(2);
      listViewWriter.startList();

      listViewWriter.bigInt().writeBigInt(0);
      listViewWriter.bigInt().writeBigInt(-127);
      listViewWriter.bigInt().writeBigInt(127);
      listViewWriter.bigInt().writeBigInt(50);
      listViewWriter.endList();

      /* write the fourth list at index 3 (empty list) */
      listViewWriter.setPosition(3);
      listViewWriter.startList();
      listViewWriter.endList();

      /* write the fifth list at index 4 */
      listViewWriter.setPosition(4);
      listViewWriter.startList();
      listViewWriter.bigInt().writeBigInt(1);
      listViewWriter.bigInt().writeBigInt(2);
      listViewWriter.bigInt().writeBigInt(3);
      listViewWriter.bigInt().writeBigInt(4);
      listViewWriter.endList();

      assertEquals(4, listViewVector.getLastSet());

      listViewVector.setValueCount(5);
}
```

Reply:
@lidavidm I updated the logic, could you please check?

Reply:
(0, 3), (NA, NA), (3, 4), (7, 0), (7, 4)

Reply:
Instead of (NA, NA), I left it (0,0) because the default values are (0,0), and I didn't do any updates.

Reply:
NA means the value doesn't matter.

Reply:
Cool. So I will keep it un updated.